util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+	} else {
+		PP_ERR("eq_readerr: %s", fi_eq_strerror(eq, eq_err.prov_errno,
+					eq_err.err_data, NULL, 0));
+
+	}
+}
+
+void pp_process_eq_err(ssize_t rd, struct fid_eq *eq, const char *fn) {
+	if (rd == -FI_EAVAIL) {
+		eq_readerr(eq);
+	} else {
+		PP_PRINTERR(fn, rd);
+	}
+}
+
+/*******************************************************************************************/
+/*                                    Addresses handling                                   */
+/*******************************************************************************************/
+
+static int pp_getsrcaddr(char *node, char *service, struct fi_info *hints)
+{
+	struct fi_info *info;
+	int ret = 0;
+
+	ret = fi_getinfo(PP_FIVERSION, node, service, FI_SOURCE, NULL, &info);
+	if (ret) {
+		PP_PRINTERR("fi_getinfo", ret);
+		return ret;
+	}
+
+	hints->src_addrlen = info->src_addrlen;
+	hints->src_addr = calloc(1, hints->src_addrlen);
+	if (!hints->src_addr) {
+		ret = -errno;
+		PP_PRINTERR("calloc", ret);
+		goto err;
+	}
+
+	memcpy(hints->src_addr, info->src_addr, hints->src_addrlen);
+	if (!hints->src_addr) {
+		ret = -errno;
+		PP_PRINTERR("memcpy", ret);
+		goto err;
+	}
+
+err:
+	fi_freeinfo(info);
+	return ret;
+}
+
+int pp_read_addr_opts(struct ct_pingpong *ct, char **node, char **service, struct fi_info *hints,
+		uint64_t *flags, struct pp_opts *opts)
+{
+	int ret;
+
+	if (opts->dst_addr) {
+		if (opts->src_addr) {
+			ret = pp_getsrcaddr(opts->src_addr, opts->src_port, hints);
+			if (ret) {
+				PP_ERR("Trying to get the source address for the client");
+				return ret;
+			}
+		}
+
+		if (!opts->dst_port)
+			opts->dst_port = ct->data_port;
+
+		*node = opts->dst_addr;
+		*service = opts->dst_port;
+	} else {
+		if (!opts->src_port)
+			opts->src_port = ct->data_port;
+
+		*node = opts->src_addr;
+		*service = opts->src_port;
+		*flags = FI_SOURCE;
+	}
+
+	return 0;
+}
+
+
+/*******************************************************************************************/
+/*                                       Test sizes                                        */
+/*******************************************************************************************/
+
+int generate_test_sizes(struct pp_opts *opts, size_t provider_maximum, int **sizes_, int size_power_two_max)
+{
+	int defaults[6] = {64, 256, 1024, 4096, 655616, 1048576};
+	int power_of_two;
+	int half_up;
+	int n = 0;
+	int *sizes = NULL;
+
+	PP_DEBUG("Generating test sizes\n");
+
+	sizes = calloc(64, sizeof(*sizes));
+	if (sizes == NULL)
+		return 0;
+	*sizes_ = sizes;
+
+	if (opts->options & PP_OPT_SIZE) {
+		if (opts->transfer_size > provider_maximum)
+			return 0;
+
+		sizes[0] = opts->transfer_size;
+		n = 1;
+	} else if (opts->sizes_enabled != PP_ENABLE_ALL) {
+		for (int i = 0; i < (sizeof defaults / sizeof defaults[0]); i++) {
+			if (defaults[i] > provider_maximum)
+				break;
+
+			sizes[i] = defaults[i];
+			n++;
+		}
+	} else {
+		for (int i = 0; i < size_power_two_max; i++) {
```

Comment:
Not necessarily 0, but 1 is usually nice to have since most pingpong tests start with 1 and some start with 0. I put 0 there because that would minimize code modification.

Fabtests can probably be changed with similar size sequence, but that can be handled separately.